src/box/memtx_tx.c

```diff
@@ -183,6 +182,200 @@ point_hole_storage_key_equal(const struct point_hole_key *key,
+	/* The transaction that will conflicted by upon commit. */
+	struct txn *victim;
+	/* Link in single-linked list. */
+	struct memtx_tx_conflict *next;
+};
+
+/**
+ * Collect an allocation to memtx_tx_stats.
+ */
+static inline void
+memtx_tx_stats_collect(struct memtx_tx_stats *stats, size_t size)
+{
+	stats->count++;
+	stats->total += size;
+}
+
+/**
+ * Discard an allocation collected by memtx_tx_stats.
+ */
+static inline void
+memtx_tx_stats_discard(struct memtx_tx_stats *stats, size_t size)
+{
+	assert(stats->count > 0);
+	assert(stats->total >= size);
+
+	stats->count--;
+	stats->total -= size;
+}
+
+/**
+ * Collect allocation statistics.
+ */
+static inline void
+memtx_tx_track_allocation(struct txn *txn, size_t size,
+			  enum memtx_tx_alloc_type alloc_type)
+{
+	assert(alloc_type < MEMTX_TX_ALLOC_MAX);
+	txn->memtx_tx_alloc_stats[alloc_type] += size;
+}
+
+/**
+ * Collect deallocation statistics.
+ */
+static inline void
+memtx_tx_track_deallocation(struct txn *txn, size_t size,
+			    enum memtx_tx_alloc_type alloc_type)
+{
+	assert(alloc_type < MEMTX_TX_ALLOC_MAX);
+	assert(txn->memtx_tx_alloc_stats[alloc_type] >= size);
+	txn->memtx_tx_alloc_stats[alloc_type] -= size;
+}
+
+/**
+ * A wrapper over mempool.
+ * Use it instead of mempool to track allocations!
+ */
+struct memtx_tx_mempool {
+	/**
+	 * Wrapped mempool.
+	 */
+	struct mempool pool;
+	/**
+	 * Each allocation is accounted with this type.
+	 */
+	enum memtx_tx_alloc_type alloc_type;
+};
+
+static inline void
+memtx_tx_mempool_create(struct memtx_tx_mempool *mempool, uint32_t objsize,
+			enum memtx_tx_alloc_type alloc_type)
+{
+	mempool_create(&mempool->pool, cord_slab_cache(), objsize);
+	mempool->alloc_type = alloc_type;
+}
+
+static inline void
+memtx_tx_mempool_destroy(struct memtx_tx_mempool *mempool)
+{
+	mempool_destroy(&mempool->pool);
+	mempool->alloc_type = MEMTX_TX_ALLOC_MAX;
+}
+
+void *
+memtx_tx_mempool_alloc(struct txn *txn, struct memtx_tx_mempool *mempool)
+{
+	void *allocation = mempool_alloc(&mempool->pool);
+	if (allocation != NULL) {
+		uint32_t size = mempool->pool.objsize;
+		memtx_tx_track_allocation(txn, size, mempool->alloc_type);
+	}
+	return allocation;
+}
+
+void
+memtx_tx_mempool_free(struct txn *txn, struct memtx_tx_mempool *mempool, void *ptr)
+{
+	uint32_t size = mempool->pool.objsize;
+	memtx_tx_track_deallocation(txn, size, mempool->alloc_type);
+	mempool_free(&mempool->pool, ptr);
+}
+
+/**
+ * Choose memtx_tx_alloc_type for alloc_obj.
+ */
+static inline enum memtx_tx_alloc_type
+memtx_tx_region_object_to_type(enum memtx_tx_alloc_object alloc_obj)
+{
+	enum memtx_tx_alloc_type alloc_type = MEMTX_TX_ALLOC_MAX;
+	switch (alloc_obj) {
+	case MEMTX_TX_OBJECT_CONFLICT:
+		alloc_type = MEMTX_TX_ALLOC_CONFLICT;
+		break;
+
+	case MEMTX_TX_OBJECT_CONFLICT_TRACKER:
+	case MEMTX_TX_OBJECT_READ_TRACKER:
+		alloc_type = MEMTX_TX_ALLOC_TRACKER;
+		break;
+	default:
+		unreachable();
+		panic("Unexpected alloc_obj in %s", __func__);
```

Comment:
AFAIC, there mustn't be anything after `unreachable`: the program is undefined at this point: https://gcc.gnu.org/onlinedocs/gcc/Other-Builtins.html.

Reply:
fixed